pytorch_memlab/line_profiler.py

```diff
@@ -58,124 +60,165 @@ def add_function(self, func):
         """ Record line profiling information for the given Python function.
         """
         try:
-            code = func.__code__
+            codehash = hash(func.__code__)
         except AttributeError:
             import warnings
             warnings.warn("Could not extract a code object for the object %r" % (func,))
             return
-        if code not in self.code_map:
-            self.code_map[code] = {}
-            self.code_map[code]['line_stat'] = defaultdict(list)
-            # probable memory leak if holding this ref
-            self.code_map[code]['func'] = func
-            self.code_map[code]['func_name'] = func.__name__
-            self.functions.append(func)
-            self.code_map[code]['source_code'] = inspect.getsourcelines(func)
-            self.code_map[code]['last_lineno'] = -1
+        if codehash not in self._codes:
+            first_line = inspect.getsourcelines(func)[1]
+            self._codes[codehash] = {
+                'func': func, 
+                'first_line': first_line,
+                'prev_line': first_line, 
+                'prev_record': -1}
```

Comment:
When profiling an function that's calling another function that's being profiled, the `reset_memory_stat`s that happen in the called function mean that the peak/accumulated memory stats in the first line after the call are wrong - they give the peak/accumulated since the last line of the inner function. To get the correct stats, it has to keep track of what the last record for the caller was, and then take maxes/sums of peaks/accumulated stats since then.